File: iHappy/iHappy/Bizhi/Control/YSERequestFetcher.m
//
//  YSERequestFetcher.m
//  iHappy
//
//  Created by dusheng.xu on 2017/5/13.
//  Copyright © 2017年 上海优蜜科技有限公司. All rights reserved.
//

#import "YSERequestFetcher.h"
#import "YSEClassifyModel.h"
#import "YSECategoryModel.h"
#import "YSEColorModel.h"
#import "YSEImageModel.h"
@interface YSERequestFetcher ()

@property (strong, nonatomic) NSDictionary *responseObj;
@property (strong, nonatomic) NSError *error;

@end

@implementation YSERequestFetcher

NSString *const kBiZhiCategoryListKey = @"categoryList";
NSString *const kBiZhiColorListKey = @"colorList";
NSString *const kBiZhiImageListKey = @"imageList";
NSString *const kBiZhiNextPageHrefKey = @"nextPageHref";
NSString *const kBiZhiIsLastPageKey = @"isLastPage";

- (void)p_fetchHomePage:(NSString *)url complete:(CompleteType)complete{
   
//    url = url.cw_URLEncodedString;
    url = [url stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    AFHTTPSessionManager *manager = [AFHTTPSessionManager manager];
    manager.responseSerializer = [AFHTTPResponseSerializer serializer];
    
    [manager GET:url parameters:nil progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSString *result = [[NSString alloc] initWithData:responseObject encoding:NSUTF8StringEncoding];
            
            TFHpple *hptts = [[TFHpple alloc] initWithHTMLData:[result dataUsingEncoding:NSUTF8StringEncoding] ];
            NSMutableArray *categoryList = [NSMutableArray arrayWithCapacity:0];
            NSMutableArray *colorList = [NSMutableArray arrayWithCapacity:0];
            NSMutableArray *imageList = [NSMutableArray arrayWithCapacity:0];
            
            NSArray *categoryItems = [hptts searchWithXPathQuery:@"//div[@class=\"options eS hidden\"]//a[@class=\"bt\"]"];
            for (TFHppleElement *element in categoryItems) {
                NSString *name = [element text];
                if ([name isEqualToString:@"动漫"] || [name isEqualToString:@"卡通"]){
                    continue;
                }
                YSECategoryModel *aModel = [[YSECategoryModel alloc] init];
                NSString *href = [element objectForKey:@"href"];
                [aModel p_setName:name href:href];
                [categoryList addObject:aModel];
            }
            
            NSArray *colorItems = [hptts searchWithXPathQuery:@"//div[@class=\"options eS hidden\"]//a[contains(@class, 'btcolor color')]"];//模糊匹配
            for (TFHppleElement *colorElement in colorItems){
                NSString *name = [colorElement text];
                NSString *href = [colorElement objectForKey:@"href"];
                YSEColorModel *colorModel = [[YSEColorModel alloc] init];
                [colorModel p_setName:name href:href];
                [colorList addObject:colorModel];
            }
            
            NSArray *imgItems = [hptts searchWithXPathQuery:@"//li[@class=\"ty-imgcont\"]//img"];//模糊匹配
            for (TFHppleElement *imgElement in imgItems){
                NSString *name = [imgElement text];
                NSString *href = [imgElement objectForKey:@"src"];
                NSString *width = [imgElement objectForKey:@"width"];
                NSString *height = [imgElement objectForKey:@"height"];

                YSEImageModel *imageModel = [[YSEImageModel alloc] init];
                [imageModel p_setTitle:name href:href width:width height:height];
                [imageList addObject:imageModel];
            }
            
            NSString *xpath_a = @"//div[@id=\"pageNum\"]//span//a";
            NSString *xpath_span = @"//div[@id=\"pageNum\"]//span//span";
            
            NSString *query = [NSString stringWithFormat:@"%@ | %@", xpath_a, xpath_span];
            NSArray *pageNumItems = [hptts searchWithXPathQuery:query];
            
            NSString *nextHref = @"";
            BOOL isLastPage = YES;
            if (pageNumItems.count > 0) {
                TFHppleElement *lastPageElement = pageNumItems.lastObject;
                nextHref = [lastPageElement objectForKey:@"href"];
                
                TFHppleElement *spanElement = pageNumItems[pageNumItems.count-2];
                isLastPage = [spanElement.tagName isEqualToString:@"span"];
            }
            
            NSDictionary *responseData = @{
                                           kBiZhiCategoryListKey:categoryList,
                                           kBiZhiColorListKey:colorList,
                                           kBiZhiImageListKey:imageList,
                                           kBiZhiNextPageHrefKey:nextHref,
                                           kBiZhiIsLastPageKey:@(isLastPage),
                                           };
            self.responseObj = responseData;
            [self safelyCallback:complete error:nil];
            
        });
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        [self safelyCallback:complete error:error];
    }];
}


- (void)safelyCallback:(CompleteType)complete error:(NSError *)error{
    self.error = error;
    if ([NSThread isMainThread]) {
        complete(self, self.responseObj, self.error);
        return;
    }
    dispatch_async(dispatch_get_main_queue(), ^{
        complete(self, self.responseObj, self.error);
    });
}

@end

File: youse壁纸站/youse/Control/YSERequestFetcher.swift
//
//  YSERequestFetcher.swift
//  youse
//
//  Created by xudosom on 16/8/27.
//  Copyright © 2016年 上海优蜜科技有限公司. All rights reserved.
//

import UIKit
typealias ResponseTuple = (requestFetcher:YSERequestFetcher, responseObj : AnyObject?, error : NSError?);
typealias CompleteType = (ResponseTuple)->Void;

let kCategoryListKey = "categoryList"
let kColorListKey = "colorList";
let kImageListKey = "imageList";
let kNextPageHrefKey = "nextPageHref";
let kIsLastPageKey = "isLastPage";

class YSERequestFetcher: NSObject {

    fileprivate var responseObj : AnyObject?;
    fileprivate var error : NSError?;

    internal func p_fetchHomePage(classifyModel:YSEClassifyModel!, complete:CompleteType) -> Void {
        let url = self.encodeEscapesURL(classifyModel.href!);
        let manager = AFHTTPSessionManager();
        manager.responseSerializer = AFHTTPResponseSerializer();
        manager.get(url, parameters: nil, progress: nil, success: { (task:URLSessionDataTask, respObject:AnyObject?) in
            
            DispatchQueue.global(priority: DispatchQueue.GlobalQueuePriority.default).async(execute: {
                let result = NSString(data: respObject as! Data, encoding: String.Encoding.utf8);
                let hptts = TFHpple.init(htmlData: result?.data(using: String.Encoding.utf8));
                
                
                var categoryList = [YSECategoryModel]();
                var colorList = [YSEColorModel]();
                var imageList = [YSEImageModel]();
                let categoryItems = hptts.search(withXPathQuery: "//div[@class=\"options eS hidden\"]//a[@class=\"bt\"]");
                for categoryElement in categoryItems{
                    let element = categoryElement as! TFHppleElement;
                    let name = element.text();
                    if name == "动漫" || name == "卡通"{
                        continue;
                    }
                    let aModel = YSECategoryModel();
                    let href = element.object(forKey: "href");
                    aModel.p_setName(name, href: href);
                    categoryList.append(aModel);
//                    NSLog("===\(element.objectForKey("href")) === \(element.text())");
                }
                
                
                let colorItems = hptts.search(withXPathQuery: "//div[@class=\"options eS hidden\"]//a[contains(@class, 'btcolor color')]");//模糊匹配
                for colorElement in colorItems{
                    let element = colorElement as! TFHppleElement;
                    let name = element.text();
                    let href = element.object(forKey: "href");
                    let colorModel = YSEColorModel();
                    colorModel.p_setName(name, href: href);
                    colorList.append(colorModel);
//                    NSLog("===\(element.objectForKey("href")) === \(element.text())");
                }

                
                let imgItems = hptts.search(withXPathQuery: "//li[@class=\"ty-imgcont\"]//img");
                for imgElement in imgItems{
                    let element = imgElement as! TFHppleElement;
                    let src = String(element.object(forKey: "src"));                    
                    let imageModel = YSEImageModel();
                    imageModel.p_setTitle(element.object(forKey: "alt"), href: src, width: element.object(forKey: "width"), height: element.object(forKey: "height"));
                    imageList.append(imageModel);
//                    NSLog("===\(src) === \(element.objectForKey("alt"))");
                }
                
                let xpath_a = "//div[@id=\"pageNum\"]//span//a";
                let xpath_span = "//div[@id=\"pageNum\"]//span//span";
                let pageNumItems = hptts.search(withXPathQuery: "\(xpath_a) | \(xpath_span)");
                let lastPageElement = pageNumItems.last as! TFHppleElement;
                let nextHref = lastPageElement.object(forKey: "href");
                
                let spanElement = pageNumItems[pageNumItems.count-2]  as! TFHppleElement;
                let isLastPage = (spanElement.tagName == "span");
                
//                for pageEle in pageNumItems{
//                    let element = pageEle as! TFHppleElement;
//                    NSLog("====nodeName = \(element.tagName) = \(element.text())");
//                }
                
                let responseData = [
                    kCategoryListKey:categoryList,
                    kColorListKey:colorList,
                    kImageListKey:imageList,
                    kNextPageHrefKey:nextHref,
                    kIsLastPageKey:isLastPage,
                ];
                self.responseObj = responseData;
                self.safelyCallback(complete, error: nil);
            });
            
        }) { (task:URLSessionDataTask?, error:NSError) in
            self.safelyCallback(complete, error: error);
        };
        
    }
    
    fileprivate func safelyCallback(_ complete:CompleteType, error:NSError?) -> Void {
        self.error = error;
        let responseTuple = (requestFetcher:self, responseObj : self.responseObj, error : self.error);
        if Thread.isMainThread {
            complete(responseTuple);
            return;
        }
        DispatchQueue.main.async { 
            complete(responseTuple);
        }
    }
    
    
    func encodeEscapesURL(_ value:String) -> String {
        let str:NSString = value
        let originalString = str as CFStringRef
        let charactersToBeEscaped = "!*'();:@&=+$,/?%#[]" as CFStringRef  //":/?&=;+!@#$()',*"    //转意符号
        //let charactersToLeaveUnescaped = "[]." as CFStringRef  //保留的符号
        let result = CFURLCreateStringByAddingPercentEscapes(kCFAllocatorDefault,
                                                             originalString,
                                                             charactersToBeEscaped,
                                                             nil,
                                                             CFStringConvertNSStringEncodingToEncoding(String.Encoding.utf8)) as NSString
        
        return result as String
    }
    
    
//    <div class="imgty" id="pageNum">
//        <span>
//                    <a class="a1">2811条</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/7.html" class="a1">上一页</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/">1</a>
//                    ..
//                    <a href="http://www.3gbizhi.com/lists-全部/4.html">4</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/5.html">5</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/6.html">6</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/7.html">7</a>
//                    <span>8</span>
//                    <a href="http://www.3gbizhi.com/lists-全部/9.html">9</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/10.html">10</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/11.html">11</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/12.html">12</a>
//                    ..
//                    <a href="http://www.3gbizhi.com/lists-全部/141.html">141</a>
//                    <a href="http://www.3gbizhi.com/lists-全部/9.html" class="a1">下一页</a>
//            </span>
//    </div>
}
